File: jevents-ui/src/types/order.ts
export interface Order {
  id: number;
  user_id: number;
  event_id: number;
  coupon_id?: number | null;
  total_amount: number;
  discount_applied: number;
  final_amount: number;
  payment_status: "pending" | "paid" | "failed" | "refunded" | "partial_refunded";
  created_at: string;
  updated_at: string;
}

export interface CreateOrderRequest {
  user_id: number;
  event_id: number;
  coupon_id?: number | null;
  total_amount: number;
  discount_applied: number;
  final_amount: number;
  payment_status: "pending"; // on creation, it's usually pending
}

export interface UpdateOrderRequest {
  coupon_id?: number | null;
  total_amount?: number;
  discount_applied?: number;
  final_amount?: number;
  payment_status?: Order["payment_status"];
}

export interface OrderResponse {
  order: Order;
}

export interface OrdersResponse {
  orders: Order[];
}

File: jevents-api/app/models/order.rb
class Order < ApplicationRecord
  belongs_to :user
  belongs_to :event
  belongs_to :coupon, optional: true

  has_many :payments, dependent: :destroy
  has_many :bookings, dependent: :nullify
  has_many :tickets, through: :bookings

  enum :payment_status, {
    pending: 0,
    paid: 1,
    failed: 2,
    refunded: 3,
    partial_refunded: 4
  }

  # validates :total_amount, :final_amount, presence: true, numericality: { greater_than_or_equal_to: 0 }
  # validates :discount_applied, numericality: { greater_than_or_equal_to: 0 }
end
